### PR TITLE
Authproxy

### PR DESCRIPTION
**Pull request by dannyroberts**
See the README and [How we will store passwords to external services](https://docs.google.com/document/d/1gH7ETG_pzi6N6hJ6h9mccGml1dJz4Hr30wc-HOjNrwk/edit) for more context.

I chose to write this in Node because javascript is my second most familiar language after python, (node http-proxy)[https://github.com/nodejitsu/node-http-proxy] seemed pretty good and worked "out of the box", while making it straightforward to intercept, reroute, change headers, etc, while I couldn't find something similar in python.

That said, my first experience writing something semi-involved in node was pretty mixed. Some plusses:

- it's super easy to, within a single process, spin up multiple http servers. This was very useful for testing, where I spin up an instance of the authproxy and an instance of a dummy server that I can proxy to for testing
- being used to ES5 and below, using ES6 is pretty nice. I especially like arrow functions that preserve `this`, `class`es, and default function argument params.

minuses:

- I'm sure I'm not using best practices, but man is this `(err, returnValue)` callback chain an annoying way to pass data up and down, and makes it super easy for errors to get swallowed by default. I know I'm supposed to use promises or something, but most of the sample code I'd search for would use this callback convention, so I just went with that.
- Some things take a long time and lots google searches to figure out how to do them. I wanted to get synchronous user input on start up, and it took me a long time to find something that worked robustly. I finally found the `deasync` library which lets you do things synchronously when you want to. (Obviously not going to write the web server code that way, but it's way simple for the start-up code to be written that way.)

### CODE REVIEW

File: services/authproxy/authproxy.js [hunk 1]
```diff
@@ -0,0 +1,129 @@
+var http = require('http'),
+    https = require('https'),
+    httpProxy = require('http-proxy'),
+    url = require('url'),
+    express = require('express'),
+    bodyParser = require('body-parser'),
+    fs = require('fs');
+
+let AUTH_PROXY_TOKEN_PASSWORD_HEADER = 'x-authproxy-token-password';
```
Comment (emord) [inline services/authproxy/authproxy.js]: Why not use `const`

**Review by emord — COMMENTED** [review]
Just glanced over this, not an in depth review

Would be interested to see what you think we could bring back to HQ/vellum after doing this. And also if you think node is more suited to some workflows we currently do in HQ.

File: services/authproxy/authproxy.js (hunk 1 above)
Comment (dannyroberts) [reply services/authproxy/authproxy.js]: don't mind if I do...

**Comment by dannyroberts** [comment]
> Would be interested to see what you think we could bring back to HQ/vellum after doing this. And also if you think node is more suited to some workflows we currently do in HQ.

Yeah I'm not sure what you have in mind exactly, but I didn't see any immediate application to CommCare HQ. The server-side code in python it seems like nothing would really justify using node for, so I assume you mean with regard to client side code and Vellum as you mentioned.

For that, I'm not really sure what to say. It'd be nice to be able to use ES6 client-side, but it would have to be useful enough to justify a transpilation step, and I don't think working on this has really given me any insight to that unfortunately.

Those are my thoughts. Let me know if I misunderstood the question.

**Comment by dannyroberts** [comment]
Since this repo's in a super-early stage and not deployed anywhere, I'm going to be pretty lax about self-merging, but will still be making PRs for running tests and visibility/feedback

**Pull request merged**